src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/ThrowsTaglet.java

```diff
@@ -47,252 +52,733 @@
+            }
+
+            @Override ThrowsTree tag() { return (ThrowsTree) super.tag(); }
+        }
+    }
+
+    /*
+     * Returns immediately inherited tags that document the provided exception type.
+     *
+     * A map associates a doc tree with its holder element externally. Such maps
+     * have defined iteration order of entries, whose keys and values
+     * are non-null.
+     */
+    private Map<ThrowsTree, ExecutableElement> expandShallowly(Element exceptionType,
+                                                               ExecutableElement holder)
+            throws Failure.ExceptionTypeNotFound,
+                   Failure.NotExceptionType,
+                   Failure.Invalid,
+                   Failure.UnsupportedTypeParameter,
+                   DocFinder.NoOverriddenMethodsFound
+    {
+        ElementKind kind = exceptionType.getKind();
+        DocFinder.Criterion<Map<ThrowsTree, ExecutableElement>, Failure> criterion;
+        if (kind == ElementKind.CLASS) {
+            criterion = method -> {
+                var tags = findByTypeElement(exceptionType, method);
+                return toResult(exceptionType, method, tags);
+            };
+        } else {
+            // type parameters are matched by position; the basis for position matching
+            // is JLS sections 8.4.2 and 8.4.4
+            criterion = method -> {
+                // TODO: add a test for the throws clause mentioning
+                //  a type parameter which is not declared by holder
+                int i = holder.getTypeParameters().indexOf((TypeParameterElement) exceptionType);
+                if (i == -1) { // the type parameter is not declared by `holder`
+                    throw new Failure.UnsupportedTypeParameter(exceptionType, null /* don't know if tag-related */, holder);
+                }
+                // if one method overrides the other, then those methods must
+                // have the same number of type parameters (JLS 8.4.2)
+                assert utils.elementUtils.overrides(holder, method, (TypeElement) holder.getEnclosingElement());
+                var typeParameterElement = method.getTypeParameters().get(i);
+                var tags = findByTypeElement(typeParameterElement, method);
+                return toResult(exceptionType, method, tags);
+            };
+        }
+        Result<Map<ThrowsTree, ExecutableElement>> result;
+        try {
+            result = utils.docFinder().trySearch(holder, criterion);
+        } catch (Failure f) {
+            // Here's why we do this ugly exception processing: the language does not allow us to
+            // instantiate the exception type parameter in criterion with a union of specific
+            // exceptions (i.e. Failure.ExceptionTypeNotFound | Failure.NotExceptionType),
+            // so we instantiate it with a general Failure. We then refine the specific
+            // exception being thrown, from the general exception we caught.
```

Comment:
In that comment I'm talking specifically about exception-type type parameters. Consider this interface:

    public interface MyInterface {

        static <X extends Throwable> void m(...) throws X { }
    }
    
Now, you can instantiate `X` like this:

    MyInterface.<EOFException>m(...);

But if you want `m` to declare that it can throw two exceptions, you cannot instantiate `X` like this:

    MyInterface.<EOFException | FileNotFoundException>m(...);

Or like this:

    MyInterface.<EOFException, FileNotFoundException>m(...);

You have to provide one exception type. It can either be a common ancestor of `EOFException` and `FileNotFoundException`; for example, `IOException`, `Exception` or `Throwable`. Or you can pack your exceptions into something else as a cause.

Whatever you choose, you won't get transparency and compile-time checking similar to what you'd get if you had this:

    static void m(...) throws FileNotFoundException, EOFException { }

Reply:
I don't (yet?) see why that language design issue arises here.   It looks like you're doing a multi-catch of a series of 4 subtypes of `Failure`, and an "other".  I don't see where any exception type parameter comes in.

Reply:
Ah OK, there's one hidden in the signature of the `trySearch` method. Is that one important?

Reply:
trySearch throws Failure, because it cannot throw a union consisting of these three exceptions using just a single exception type parameter:

* Failure.UnsupportedTypeParameter
* Failure.ExceptionTypeNotFound
* Failure.NotExceptionType

So I make a general "Failure" mixture, from which I'll later refine individual constituents. I think we can do this, but that's it:

        try {
            result = utils.docFinder().trySearch(holder, criterion);
        } catch (Failure.NotExceptionType
                 | Failure.ExceptionTypeNotFound
                 | Failure.UnsupportedTypeParameter x) {
            throw x;
        } catch (Failure f) {
            throw newAssertionError(f);
        }

Maybe my comment was misplaced (the comment is about `Criterion`, not `trySearch`)? I can perhaps remove it altogether if it doesn't bring much value.

Reply:
👍 for the immediately preceding code sample for a simplified `try`